### PR TITLE
[16.0][ADD] sale_order_warehouse_from_delivery_carrier

### PR DESCRIPTION
**Pull request by geomer198**
This module allows to set warehouse in Sales Order based on selected shipping method.

### CODE REVIEW

**Review by ivs-cetmix — APPROVED** [review]
Functional ok

File: sale_order_wh_from_delivery_carrier/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.depends("user_id", "company_id", "carrier_id")
+    def _compute_warehouse_id(self):
+        # Set warehouse by shipping method
+        so_with_shipping = self.filtered("carrier_id")
+        for order in so_with_shipping:
+            if order.state == "draft" and order.carrier_id.so_warehouse_id:
```
Comment (GabbasovDinar) [inline sale_order_wh_from_delivery_carrier/models/sale_order.py]: computer method uses in its calculations also the `state` field of the order, please add it to api.depends

Comment (geomer198) [reply sale_order_wh_from_delivery_carrier/models/sale_order.py]: Resolved

**Review by GabbasovDinar — APPROVED** [review]
LGTM

**Comment by ivs-cetmix** [comment]
Hey @OCA/logistics-maintainers could you please have a look at this?

**Comment by rousseldenis** [comment]
@geomer198 Thanks for this. I would say I'd prefer explicit words in modules names as they can quickly seems like module x_y_z which does not mean anything.

**Comment by ivs-cetmix** [comment]
> @geomer198 Thanks for this. I would say I'd prefer explicit words in modules names as they can quickly seems like module x_y_z which does not mean anything.

Thank you for the feedback. Any ideas how this module name should look like?

File: sale_order_wh_from_delivery_carrier/models/delivery_carrier.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from odoo import fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    so_warehouse_id = fields.Many2one(
+        comodel_name="stock.warehouse", string="SO Warehouse"
```
Comment (rousseldenis) [inline sale_order_wh_from_delivery_carrier/models/delivery_carrier.py]: A help is welcome here.

File: sale_order_wh_from_delivery_carrier/tests/test_module_flow.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+from odoo.tests import Form, TransactionCase, tagged
+
+
+@tagged("post_install", "-at_install")
+class TestModuleFlow(TransactionCase):
+    def setUp(self):
```
Comment (rousseldenis) [inline sale_order_wh_from_delivery_carrier/tests/test_module_flow.py]: Use classmethod instead.

File: test-requirements.txt [hunk 4]
```diff
@@ -1,3 +1,5 @@
 odoo-test-helper
 vcrpy
 vcrpy-unittest
+odoo-addon-delivery_roulier @ git+https://github.com/OCA/delivery-carrier.git@refs/pull/731/head#subdirectory=setup/delivery_roulier
```
Comment (rousseldenis) [inline test-requirements.txt]: @geomer198 Why adding these as they are not in this module dependencies ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]

File: test-requirements.txt (hunk 4 above)
Comment (geomer198) [reply test-requirements.txt]: Added for fixing broken tests in delivery_roulier module.

**Comment by geomer198** [comment]
@rousseldenis Could you please check my changes?

**Comment by geomer198** [comment]
@rousseldenis Could you check changes?

**Review by rousseldenis — APPROVED** [review]
Code review

File: sale_order_wh_from_delivery_carrier/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+{
+    "name": "Sale Order WH from Delivery Carrier",
```
Comment (jbaudoux) [inline sale_order_wh_from_delivery_carrier/__manifest__.py]:
```suggestion
    "name": "Sales Order Warehouse from Delivery Carrier",
```

File: sale_order_wh_from_delivery_carrier/models/delivery_carrier.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from odoo import fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    so_warehouse_id = fields.Many2one(
+        comodel_name="stock.warehouse",
+        string="SO Warehouse",
```
Comment (jbaudoux) [inline sale_order_wh_from_delivery_carrier/models/delivery_carrier.py]:
```suggestion
        string="Sales Order Warehouse",
```

File: sale_order_wh_from_delivery_carrier/models/delivery_carrier.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+        help="Default Warehouse for sale where uses current Delivery Carrier",
```
Comment (jbaudoux) [inline sale_order_wh_from_delivery_carrier/models/delivery_carrier.py]:
```suggestion
        help="Default warehouse for sales having this delivery carrier",
```

File: sale_order_wh_from_delivery_carrier/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+from odoo import api, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.depends(
+        "user_id", "company_id", "state", "carrier_id", "carrier_id.so_warehouse_id"
```
Comment (jbaudoux) [inline sale_order_wh_from_delivery_carrier/models/sale_order.py]: You don't need to trigger recomputation on state change or if you reconfigure the carrier
```suggestion
        "user_id", "company_id", "carrier_id"
```

**Review by jbaudoux — COMMENTED** [review]
Pls prev. to use abbr. Thx :)
wh => warehouse

I would have named the module sale_warehouse_carrier or sale_order_warehouse_carrier
and put that in sale-workflow repo

**Comment by rousseldenis** [comment]
> > @geomer198 Thanks for this. I would say I'd prefer explicit words in modules names as they can quickly seems like module x_y_z which does not mean anything.
> 
> Thank you for the feedback. Any ideas how this module name should look like?

`sale_order_warehouse_from_delivery_carrier`

**Comment by geomer198** [comment]
> > > @geomer198 Thanks for this. I would say I'd prefer explicit words in modules names as they can quickly seems like module x_y_z which does not mean anything.
> > 
> > 
> > Thank you for the feedback. Any ideas how this module name should look like?
> 
> `sale_order_warehouse_from_delivery_carrier`

Module is renamed. Could you check changes?

File: sale_order_warehouse_from_delivery_carrier/models/delivery_carrier.py [hunk 9]
```diff
@@ -0,0 +1,11 @@
+from odoo import fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    so_warehouse_id = fields.Many2one(
+        comodel_name="stock.warehouse",
+        string="SO Warehouse",
+        help="Default Warehouse for sale where uses current Delivery Carrier",
```
Comment (jbaudoux) [inline sale_order_warehouse_from_delivery_carrier/models/delivery_carrier.py]:
```suggestion
        help="Default warehouse for the sales order that uses this delivery carrier",
```

**Review by jbaudoux — APPROVED** [review]

**Review by rousseldenis — APPROVED** [review]

File: sale_order_warehouse_from_delivery_carrier/models/sale_order.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+from odoo import api, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.depends(
+        "user_id", "company_id", "state", "carrier_id", "carrier_id.so_warehouse_id"
```
Comment (jbaudoux) [inline sale_order_warehouse_from_delivery_carrier/models/sale_order.py]: This wasn't solved. I don't see a reason for recomputation on state change. And you don't want to recompute all your SO if the config on the carrier is changed
```suggestion
        "user_id", "company_id", "carrier_id"
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]

File: sale_order_warehouse_from_delivery_carrier/models/delivery_carrier.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
+from odoo import fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = "delivery.carrier"
+
+    so_warehouse_id = fields.Many2one(
+        comodel_name="stock.warehouse",
+        string="SO Warehouse",
```
Comment (jbaudoux) [inline sale_order_warehouse_from_delivery_carrier/models/delivery_carrier.py]:
```suggestion
        string="Sales Order Warehouse",
```

File: sale_order_warehouse_from_delivery_carrier/models/sale_order.py (hunk 10 above)
Comment (rousseldenis) [reply sale_order_warehouse_from_delivery_carrier/models/sale_order.py]: Agree, didn't see

File: sale_order_warehouse_from_delivery_carrier/readme/CONFIGURE.rst [hunk 12]
```diff
@@ -0,0 +1,3 @@
+Go to `Sales` -> `Settings` -> `Shipping methods` and select a shipping method
+
+Set warehouse in the "SO Warehouse" field.
```
Comment (jbaudoux) [inline sale_order_warehouse_from_delivery_carrier/readme/CONFIGURE.rst]:
```suggestion
Set warehouse in the "Sales Order Warehouse" field.
```

File: sale_order_warehouse_from_delivery_carrier/readme/USAGE.rst [hunk 13]
```diff
@@ -0,0 +1,7 @@
+Create a new quotation and select a shipping method.
+
+If selected shipping method has a SO warehouse defined
```
Comment (jbaudoux) [inline sale_order_warehouse_from_delivery_carrier/readme/USAGE.rst]:
```suggestion
If selected shipping method has a sales order warehouse defined
```

File: sale_order_warehouse_from_delivery_carrier/readme/USAGE.rst [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+this warehouse will be automatically set in the quotation.
+
+**Note**:
+warehouse will be NOT updated if Sales Order is already confirmed.
```
Comment (jbaudoux) [inline sale_order_warehouse_from_delivery_carrier/readme/USAGE.rst]:
```suggestion
The warehouse will NOT be updated if the sales order is already confirmed.
```

**Comment by geomer198** [comment]
@rousseldenis, @jbaudoux  Could you check again?

**Review by jbaudoux — APPROVED** [review]

**Comment by jbaudoux** [comment]
> @rousseldenis, @jbaudoux Could you check again?

Thanks.

/ocabot merge nobump

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
You need to add `DESCRIPTION.rst`

**Comment by jbaudoux** [comment]
@geomer198 can you squash the commits?

**Review by ivs-cetmix — APPROVED** [review]
LGTM

**Pull request merged**